include/rocksdb/write_buffer_manager.h

```diff
@@ -56,121 +64,129 @@ class WriteBufferManager final {
+        IsStallThresholdExceeded()) {
+      return true;
+    } else {
       return false;
     }
-
-    return IsStallActive() || IsStallThresholdExceeded();
-  }
-
-  // Returns true if stall is active.
-  bool IsStallActive() const {
-    return stall_active_.load(std::memory_order_relaxed);
-  }
-
-  // Returns true if stalling condition is met.
-  bool IsStallThresholdExceeded() const {
-    return memory_usage() >= buffer_size_;
   }
 
+  // WARNING: Should only be called by RocksDB internally.
   void ReserveMem(size_t mem);
 
   // We are in the process of freeing `mem` bytes, so it is not considered
   // when checking the soft limit.
+  //
+  // WARNING: Should only be called by RocksDB internally.
   void ScheduleFreeMem(size_t mem);
 
+  // WARNING: Should only be called by RocksDB internally.
   void FreeMem(size_t mem);
 
-  // Add the DB instance to the queue and block the DB.
-  // Should only be called by RocksDB internally.
-  void BeginWriteStall(StallInterface* wbm_stall);
+  // Return true if stall conditions have been met and WriteBufferManager has
+  // done its work to begin write stall. Return false otherwise.
+  //
+  // WARNING: Should only be called by RocksDB internally.
+  bool MaybeBeginWriteStall(StallInterface* wbm_stall);
+
+  // WARNING: Should only be called by RocksDB internally.
+  void RemoveDBFromQueue(StallInterface* wbm_stall);
 
+ private:
   // If stall conditions have resolved, remove DB instances from queue and
   // signal them to continue.
+  //
+  // Called when stall conditions (see `ShouldStall()` API) are changed
+  //
+  // REQUIRED: wbm_mutex_ held
   void MaybeEndWriteStall();
 
-  void RemoveDBFromQueue(StallInterface* wbm_stall);
+  // REQUIRED: wbm_mutex_ held
+  bool IsStallThresholdExceeded() const {
+    return memory_usage() >= buffer_size_;
+  }
+
+  // Returns the total memory used by active memtables.
+  //
+  // REQUIRED: wbm_mutex_ held
+  size_t mutable_memtable_memory_usage() const {
+    return memory_active_.load(std::memory_order_relaxed);
+  }
+
+  // WARNING:  Should only be called from write thread
+  // REQUIRED: wbm_mutex_ held
+  void ReserveMemWithCache(size_t mem);
+
+  // REQUIRED: wbm_mutex_ held
+  void FreeMemWithCache(size_t mem);
+
+  // Mutex used to protect WriteBufferManager's data variables.
+  mutable std::mutex wbm_mutex_;
 
- private:
   std::atomic<size_t> buffer_size_;
   std::atomic<size_t> mutable_limit_;
   std::atomic<size_t> memory_used_;
   // Memory that hasn't been scheduled to free.
   std::atomic<size_t> memory_active_;
   std::shared_ptr<CacheReservationManager> cache_res_mgr_;
-  // Protects cache_res_mgr_
-  std::mutex cache_res_mgr_mu_;
 
   std::list<StallInterface*> queue_;
-  // Protects the queue_ and stall_active_.
-  std::mutex mu_;
-  bool allow_stall_;
-  // Value should only be changed by BeginWriteStall() and MaybeEndWriteStall()
-  // while holding mu_, but it can be read without a lock.
-  std::atomic<bool> stall_active_;
```

Comment:
Cleanup2: I don't see why stall_active_ is needed. As long as we have our stall condition check (ShouldStall()) implemented right, stall_active_ == ShouldStall(). But let me know if I overlook anything :)

Reply:
@akankshamahajan15 shared an perspective of `stall_active_` might exists as a perf optimization to reduce lock contention for the case of multiple DB using same WBM. She will cite more previous discussion on this soon. 

I am not entirely sure about keeping this yet mainly for the reason that such perf optimization makes the concurrency model of WBM harder to understand as some can be access without lock while some can't.

[TODO for me] Understand the previous conversation on having stall_active_ the need of that; reconsider again the perf cost VS model simplicity